airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialReadUtil.java

```diff
@@ -171,13 +174,38 @@ public static List<AutoCloseableIterator<AirbyteMessage>> getCdcReadIterators(fi
             AirbyteTraceMessageUtility::emitStreamStatusTrace));
   }
 
+  public static Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyLoadStatus> initPairToPrimaryKeyLoadStatusMap(
+                                                                                                                                       final Map<io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair, PrimaryKeyLoadStatus> pairToPkStatus) {
+    final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyLoadStatus> map = new HashMap<>();
+    pairToPkStatus.forEach((pair, pkStatus) -> {
+      final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair updatedPair =
+          new io.airbyte.protocol.models.AirbyteStreamNameNamespacePair(pair.getName(), pair.getNamespace());
+      map.put(updatedPair, pkStatus);
+    });
+    return map;
+  }
+
+  public static Set<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair> initStreamsCompletedSnapshot(final InitialLoadStreams initialLoadStreams,
+                                                                                                            final ConfiguredAirbyteCatalog catalog) {
+    final Set<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair> streamsThatHaveCompletedSnapshot = new HashSet<>();
+    catalog.getStreams().forEach(configuredAirbyteStream -> {
+      if (!initialLoadStreams.streamsForInitialLoad().contains(configuredAirbyteStream)
+          && configuredAirbyteStream.getSyncMode() == SyncMode.INCREMENTAL) {
+        streamsThatHaveCompletedSnapshot.add(
+            new io.airbyte.protocol.models.AirbyteStreamNameNamespacePair(configuredAirbyteStream.getStream().getName(),
+                configuredAirbyteStream.getStream().getNamespace()));
+      }
+    });
+    return streamsThatHaveCompletedSnapshot;
+  }
+
   /**
-   * Determines the streams to sync for initial primary key load. These include streams that are (i)
-   * currently in primary key load (ii) newly added incremental streams.
+   * CDC specific: Determines the streams to sync for initial primary key load. These include streams
+   * that are (i) currently in primary key load (ii) newly added incremental streams.
    */
-  public static InitialLoadStreams streamsForInitialPrimaryKeyLoad(final CdcStateManager stateManager,
-                                                                   final ConfiguredAirbyteCatalog fullCatalog,
-                                                                   final boolean savedOffsetStillPresentOnServer) {
+  public static InitialLoadStreams cdcStreamsForInitialPrimaryKeyLoad(final CdcStateManager stateManager,
```

Comment:
The methods `cdcStreamsForInitialPrimaryKeyLoad` and  `streamsForInitialPrimaryKeyLoad` have a lot in common except for the offset check. Can be refactored into a single method.